src/core/mac/mac_frame.cpp

```diff
@@ -836,75 +810,71 @@ uint8_t Frame::CalculateAddrFieldSize(uint16_t aFcf)
 
 uint8_t Frame::FindPayloadIndex(void) const
 {
-    uint8_t index = SkipSecurityHeaderIndex();
+    // We use `uint16_t` for `index` to handle its potential roll-over
+    // while parsing and verifying Header IE(s).
+
+    uint16_t index = SkipSecurityHeaderIndex();
 
     VerifyOrExit(index != kInvalidIndex, OT_NOOP);
 
 #if OPENTHREAD_CONFIG_MAC_HEADER_IE_SUPPORT
     if (IsIePresent())
     {
-        const uint8_t *cur    = &mPsdu[index];
-        const uint8_t *footer = GetFooter();
+        uint8_t footerLength = GetFooterLength();
 
         do
         {
-            const HeaderIe *ie;
-            uint8_t         len;
+            const HeaderIe *ie = reinterpret_cast<const HeaderIe *>(&mPsdu[index]);
 
-            VerifyOrExit(cur + sizeof(HeaderIe) <= footer, index = kInvalidIndex);
-
-            ie  = reinterpret_cast<const HeaderIe *>(cur);
-            len = static_cast<uint8_t>(ie->GetLength());
-
-            cur += sizeof(HeaderIe);
             index += sizeof(HeaderIe);
 
-            VerifyOrExit(cur + len <= footer, index = kInvalidIndex);
+            VerifyOrExit(index + footerLength <= mLength, index = kInvalidIndex);
 
-            cur += len;
-            index += len;
+            index += ie->GetLength();
+            VerifyOrExit(index + footerLength <= mLength, index = kInvalidIndex);
```

Comment:
While I do like the idea of this optimization and it seems technically correct (since adding an extra non-negative `ie->GetLength()` should only make the`VerifyOrExit` condition check more strict), I am not sure if it is safe and fine to do this.

Without the first check, the call to `ie->GetLength()` can potentially read and access beyond the length and end of the `mPsdu` frame buffer. Since the buffer is provided by platform layer code, within OT core we cannot know for sure where the `mPsdu` buffer is actually pointing to and should avoid reading beyond its given length.

I think some our CI checks (if I recall correctly the fuzzer check) would try to detect access/read beyond buffer length and may flag it as a bug/issue.

Reply:
Make sense. We should keep it safer.